genesyscloud/resource_genesyscloud_integration.go

```diff
@@ -0,0 +1,495 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/mypurecloud/platform-client-sdk-go/platformclientv2"
+)
+
+var (
+	integrationConfigResource = &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "Integration config name.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"notes": {
+				Description: "Integration config notes.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"properties": {
+				Description: "Integration config properties.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+			},
+			"advanced": {
+				Description: "Integration advanced config.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+			},
+			"credentials": {
+				Description: "Credentials required for the integration. The required keys are indicated in the credentials property of the Integration Type.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+			},
+		},
+	}
+)
+
+func getAllIntegrations(ctx context.Context, clientConfig *platformclientv2.Configuration) (ResourceIDMetaMap, diag.Diagnostics) {
+	resources := make(ResourceIDMetaMap)
+	integrationAPI := platformclientv2.NewIntegrationsApiWithConfig(clientConfig)
+
+	for pageNum := 1; ; pageNum++ {
+		integrations, _, err := integrationAPI.GetIntegrations(100, pageNum, "", nil, "", "")
+		if err != nil {
+			return nil, diag.Errorf("Failed to get page of integrations: %v", err)
+		}
+
+		if integrations.Entities == nil || len(*integrations.Entities) == 0 {
+			break
+		}
+
+		for _, integration := range *integrations.Entities {
+			resources[*integration.Id] = &ResourceMeta{Name: *integration.Name}
+		}
+	}
+
+	return resources, nil
+}
+
+func integrationExporter() *ResourceExporter {
+	return &ResourceExporter{
+		GetResourcesFunc: getAllWithPooledClient(getAllIntegrations),
+		RefAttrs: map[string]*RefAttrSettings{
+			// TODO: Since now it uses jsonencode, might need to change how export works
+			"config.credentials.*": {RefType: "genesyscloud_integration_credentials"},
+			"integration_type":     {},
+		},
+	}
+}
+
+func resourceIntegration() *schema.Resource {
+	return &schema.Resource{
+		Description: "Genesys Cloud Integration",
+
+		CreateContext: createWithPooledClient(createIntegration),
+		ReadContext:   readWithPooledClient(readIntegration),
+		UpdateContext: updateWithPooledClient(updateIntegration),
+		DeleteContext: deleteWithPooledClient(deleteIntegration),
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		SchemaVersion: 1,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "Integration name.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"intended_state": {
```

Comment:
Can we define the options for intended_state in the description and add a validator to make sure the config specifies one of them?

Reply:
Added options and validate function

Reply:
Can you actually set the integration state to "DELETED"? I see it listed in the API doc, but it may be a state that would be returned from a GET after you've called DELETE. Worth trying out to see if we should include it here.